<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.5"/>
<title>btllib: Main Page</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">btllib
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.5 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li class="current"><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li><a href="examples.html"><span>Examples</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
</div><!-- top -->
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Variables</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="headertitle">
<div class="title">btllib Documentation</div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p><a href="https://github.com/bcgsc/btllib/releases/latest">![GitHub release (latest by date)](https://img.shields.io/github/v/release/bcgsc/btllib?label=Release)</a> <a href="https://anaconda.org/bioconda/btllib">![Conda](https://img.shields.io/conda/dn/bioconda/btllib?label=Conda%20downloads)</a> <a href="https://dev.azure.com/bcgsc/btl_public/_build/latest?definitionId=1">![Build Status](https://dev.azure.com/bcgsc/btl_public/_apis/build/status/bcgsc.btllib)</a> <a href="https://doi.org/10.21105/joss.04720">![DOI](https://joss.theoj.org/papers/10.21105/joss.04720/status.svg)</a></p>
<p><a href="http://www.birollab.ca/">Bioinformatics Technology Lab</a> common code library in C++ with Python wrappers.</p>
<h2>Platforms </h2>
<ul>
<li>Linux</li>
<li>MacOS</li>
</ul>
<h2>Installation for users </h2>
<p>The recommended way is to download using <a href="https://docs.conda.io/en/latest/">Conda</a> package manager: <code>conda install -c bioconda -c conda-forge btllib</code></p>
<p>Alternatively, you can compile the code from source. Download <code>btllib-$VERSION.tar.gz</code> from the GitHub <a href="https://github.com/bcgsc/btllib/releases/latest">latest release</a> where <code>$VERSION</code> is the latest btllib version and do the following:</p>
<ul>
<li><code>tar xzf btllib-$VERSION.tar.gz</code> to extract the source code.</li>
<li>Have the dependencies ready:<ul>
<li>GCC 6+ or Clang 5+ (with OpenMP and C++17 support)</li>
<li>Python 3.9+</li>
<li>Meson and Ninja Python3 packages, CMake (If not available, these will be automatically installed to a temporary directory.)</li>
</ul>
</li>
<li>Run <code>btllib/compile</code><ul>
<li>This will install btllib in the <code>btllib/install</code> directory. You can provide the <code>--prefix</code> parameter to change this.</li>
<li>The C++ compiler must be the same as the one used for compiling Python. E.g. if you installed Python using a package manager, you should use the C++ compiler from the same package manager. You can change the compiler by exporting the <code>CXX</code> environment variable to point to the compiler before running <code>btllib/compile</code>.</li>
<li>You can optionally run <code>python3 -m pip install $PREFIX/lib/btllib/python</code> afterwards to install the Python package. The Python wrappers are usable even without this step. <code>$PREFIX</code> is the path where btllib is installed.</li>
</ul>
</li>
</ul>
<h2>Using the library </h2>
<ul>
<li>Run time dependencies:<ul>
<li>SAMtools for reading SAM, BAM, and CRAM files.</li>
<li>gzip, tar, pigz, bzip2, xz, lrzip, zip, and/or 7zip for compressing/decompressing files. Not all of these are necessary, only the ones whose compressions you'll be using.</li>
<li>wget for downloading sequences from a URL.</li>
</ul>
</li>
<li>Building C++ code (<code>$PREFIX</code> is the path where btllib is installed):<ul>
<li>Link your code with <code>$PREFIX/lib/libbtllib.a</code> (pass <code>-L $PREFIX/lib -l btllib</code> flags to the compiler).<ul>
<li>You can do so by typing the following in your console:<ul>
<li><code>export CPPFLAGS="-isystem /path/to/btllib/install/include $CPPFLAGS"</code></li>
<li><code>export LDFLAGS="-L/path/to/btllib/install//lib -lbtllib $LDFLAGS"</code></li>
</ul>
</li>
</ul>
</li>
<li><code>#include</code> any header from the <code>$PREFIX/include</code> directory (pass <code>-I $PREFIX/include</code> flag to the compiler).</li>
<li><code>btllib</code> uses <code>C++11</code> features, so that standard should be enabled at a minimum.</li>
</ul>
</li>
<li>Running Python code:<ul>
<li>The Python used to import btllib <em>must</em> be the same as the one used to compile the library. Specifically, btllib uses <code>python3-config</code> to determine the flags used for compilation. Running <code>python3-config --exec-prefix</code> will give the path to the Python installation that needs to be used. The <code>python3</code> executable can be found at <code>$(python3-config --exec-prefix)/bin/python3</code>.</li>
<li>The wrappers correspond one-to-one with C++ code so any functions and classes can be used under the same name. The only exceptions are nested classes which are prefixed with outer class name (e.g. <code><a class="el" href="structbtllib_1_1SeqReader_1_1Flag.html">btllib::SeqReader::Flag</a></code> in C++ versus <code>btllib.SeqReaderFlag</code> in Python), and (Kmer)CountingBloomFilter which provides <code>CountingBloomFilter8</code>, <code>CountingBloomFilter16</code>, <code>CountingBloomFilter32</code>, <code>KmerCountingBloomFilter8</code>, <code>KmerCountingBloomFilter16</code>, <code>CountingBloomFilter32</code> with counters 8, 16, and 32 bits wide.</li>
<li>If you compiled btllib from source code and didn't install the Python wrappers, you can use <code>PYTHONPATH</code> environment variable or <code>sys.path.append()</code> in your Python code to include <code>$PREFIX/lib/btllib/python/btllib</code> directory to make btllib available to the interpreter.</li>
<li>Include the library with <code>import btllib</code></li>
</ul>
</li>
<li>Executables<ul>
<li>btllib generated executables can be found in <code>$PREFIX/bin</code> directory. Append that path to the <code>PATH</code> environment variable to make it available to your shell.</li>
</ul>
</li>
</ul>
<h2>Documentation </h2>
<p><a href="https://bcgsc.github.io/btllib/">Docs page</a></p>
<h2>For btllib developers </h2>
<ul>
<li>Initial setup:<ul>
<li><code>git clone --recurse-submodules <a href="https://github.com/bcgsc/btllib">https://github.com/bcgsc/btllib</a></code> in order to obtain all the code.</li>
<li>In <code>btllib</code> dir, run <code>meson build</code> to create a build directory.</li>
</ul>
</li>
<li>Every time you want to run tests, in the <code>build</code> dir:<ul>
<li><code>ninja wrap</code> to regenerate wrappers.</li>
<li><code>ninja test</code> to build wrappers and tests, and run tests.</li>
</ul>
</li>
<li>Before making a pull request, in the <code>build</code> dir:<ul>
<li><code>ninja quality-assurance</code> to make sure all CI tests pass.</li>
<li>Make a commit after the above step, in case it has made any changes to wrappers or formatting. Don't commit the changes made to the <code>sdsl-lite</code> subproject. Meson config file adjusts the <code>sdsl-lite</code> config in order for it to work for <code>btllib</code>, but this is done ad hoc and is not necessary to be committed. By doing it ad hoc we keep a list of differences compared to the upstream repository.</li>
</ul>
</li>
<li>Before making a release, in the <code>build</code> dir:<ul>
<li>Do the same as for a pull request and</li>
<li><code>ninja docs</code> to regenerate docs to reflect the release and then commit the changes.</li>
<li><code>meson dist --allow-dirty</code> to generate a self-contained package based on the last commit. <code>--allow-dirty</code> permits making a distributable with uncommited changes. This is necessary as <code>sdsl-lite</code> dependency has ad hoc changes made during the build process. The resulting distributable will be compressed with xz. For easier use, decompress it and then compress with gzip. Attach the resulting file to the release.</li>
</ul>
</li>
</ul>
<p>The following are all the available <code>ninja</code> commands which can be run within <code>build</code> directory:</p>
<ul>
<li><code>ninja clang-format</code> formats the whitespace in code (requires clang-format 8+).</li>
<li><code>ninja wrap</code> wraps C++ code for Python (requires SWIG 4.0+).</li>
<li><code>ninja clang-tidy</code> runs clang-tidy on C++ code and makes sure it passes (requires clang-tidy 8+).</li>
<li><code>ninja</code> builds the tests and wrapper libraries / makes sure they compile.</li>
<li><code>ninja test</code> runs the tests.</li>
<li><code>ninja code-coverage</code> assures code coverage threshold is satisfied. (requires gcovr 3.3+)</li>
<li><code>ninja sanitize-undefined</code> runs undefined sanitization.</li>
<li><code>ninja test-wrappers</code> tests whether wrappers work.</li>
<li><code>ninja docs</code> generates code documentation from comments (requires Doxygen).</li>
<li><code>ninja quality-assurance</code> runs <code>clang-format</code>, <code>wrap</code>, <code>clang-tidy</code>, <code>test</code>, <code>code-coverage</code>, <code>sanitize-undefined</code>, and <code>test-wrappers</code>. These are all checked at the CI test.</li>
</ul>
<h2>Credits </h2>
<ul>
<li>Author: <a href="https://github.com/vlad0x00">Vladimir Nikolic</a></li>
<li>Components:<ul>
<li><a href="https://github.com/mohamadi">Hamid Mohamadi</a> and <a href="https://github.com/parham-k">Parham Kazemi</a> for <a href="https://github.com/bcgsc/ntHash">ntHash</a></li>
<li><a href="https://github.com/JustinChu">Justin Chu</a> for <a href="https://github.com/bcgsc/btl_bloomfilter">MIBloomFilter</a></li>
</ul>
</li>
<li>Included dependencies:<ul>
<li><a href="https://github.com/skystrife">Chase Geigle</a> for <a href="https://github.com/skystrife/cpptoml">cpptoml</a></li>
<li>Simon Gog, Timo Beller, Alistair Moffat, and Matthias Petri for <a href="https://github.com/simongog/sdsl-lite">sdsl-lite</a></li>
</ul>
</li>
</ul>
<h2>Citing </h2>
<p>If you use btllib in your research, please cite:</p>
<p>NikoliÄ‡ et al., (2022). btllib: A C++ library with Python interface for efficient genomic sequence processing. Journal of Open Source Software, 7(79), 4720, <a href="https://doi.org/10.21105/joss.04720">https://doi.org/10.21105/joss.04720</a></p>
<p>If you use aaHash in your research, please cite:</p>
<p>Wong et al., (2023). aaHash: recursive amino acid sequence hashing. Bioinformatics Advances, vbad162, <a href="https://doi.org/10.1093/bioadv/vbad162">https://doi.org/10.1093/bioadv/vbad162</a>. </p>
</div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.5
</small></address>
</body>
</html>
